storage: Add common amounts commitment lookup table

SyntheticBird45 opened the pull request:
<!--
PR titles should be:
<AREA>: <SHORT_DESCRIPTION>

For example:
books: fix typo
-->

<!--
If your pull request is long and/or has sections
that need clarifying, consider leaving a review on
your own PR with comments explaining the changes.
-->

### What
<!--
If applicable, close a related issue with: 

Fixes #<BUG_ISSUE_NUMBER>

...or...

Closes #<FEATURE_ISSUE_NUMBER>
-->

Fixes #245

<!-- Describe the pull request in detail. -->
Implements monerod's `zero_commitment` lookup table for common amounts defined here: https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/ringct/rctOps.cpp#L43

Unlike monerod which uses a simple linear search, this PR implements a binary search over the lookup table, saving a few dozen nanoseconds. 

More micro-optimization could be done like verifying if the amount is over 10 then modulo of amount per 10 should be 0, saving ourselves the lookup cost by directly passing to calculation.

Boog900 commented on constants/src/crypto.rs:
```diff
@@ -0,0 +1,184 @@
+/// Defined at:
+/// - <https://github.com/monero-project/monero/blob/893916ad091a92e765ce3241b94e706ad012b62a/src/ringct/rctOps.cpp#L44>
+#[rustfmt::skip]
+pub static ZERO_COMMITMENT_LOOKUP_TABLE: [(u64, [u8;32]); 173] = 
+[
+  (0_u64,                       [0x58, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66] ),
```
Can we calcualte these at runtime using a LazyLock, like I did here: https://github.com/Cuprate/cuprate/blob/1eaf32bee956b66591f1537201c16969d5ec8696/helper/src/commitment.rs#L55-L62

Boog900 commented on storage/blockchain/src/ops/output.rs:
```diff
@@ -155,9 +158,31 @@ pub fn output_to_output_on_chain(
-    // FIXME: implement lookup table for common values:
-    // <https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/ringct/rctOps.cpp#L322>
-    let commitment = ED25519_BASEPOINT_POINT + *H * Scalar::from(amount);
+    // Binary search in lookup table (O(log2(n)+2) instead of O(n+2))
+    let commitment: EdwardsPoint = {
+        let (mut start, mut end, mut c): (usize, usize, Option<EdwardsPoint>) =
+            (0, (ZC_LT.len() - 1) / 2, None);
+        while start <= end {
+            let mid = (start + end) / 2;
+            let lookup = ZC_LT[mid].0;
+            match amount {
+                a if a == lookup => {
+                    c = CompressedEdwardsY::from_slice(&ZC_LT[mid].1)
+                        .expect("The lookup table cannot contain invalid curve point data")
+                        .decompress();
+                }
+                a if a < lookup => {
+                    end = mid - 1;
+                }
+                _ => {
+                    start = mid + 1;
+                }
+            }
+        }
+
+        // Compute the zero commitment if not found.
+        c.unwrap_or_else(|| ED25519_BASEPOINT_POINT + *H * Scalar::from(amount))
```
Can we have this function with the commitment table please

Boog900 commented on storage/blockchain/src/ops/output.rs:
```diff
@@ -155,9 +158,31 @@ pub fn output_to_output_on_chain(
+            let lookup = ZC_LT[mid].0;
+            match amount {
+                a if a == lookup => {
+                    c = CompressedEdwardsY::from_slice(&ZC_LT[mid].1)
+                        .expect("The lookup table cannot contain invalid curve point data")
+                        .decompress();
```
Doing a scalar decomposition here is pretty expensive - the cache could be the decompressed EdwardsPoints

Boog900 replied:
We could keep the decomposed point in the cache instead of doing a decompression on every use.

SyntheticBird45 commented:
Applied requested changes however I wanted to benchmark hashmap for the lookup. The benchmark is jut a lookup over all the decomposed amounts. 
- I kept manual binary search
- Implemented the opportunistic optimization with the modulo. If the value is over 10 and the modulo is different than 0 then it skip the lookup as we now the value isn't in the table.
- Since `cuprate-constants` do not have `std` support I moved back the code to `cuprate-helper` similarly to your test branch.

Benchmark:

![benchmark_hashmap](https://github.com/user-attachments/assets/059c3e47-97d1-4735-8dc3-baeebb8a33df)

Boog900 commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,96 @@
+        let mid = (start + end) / 2;
+        let lookup = ZERO_COMMITMENT_LOOKUP_TABLE[mid].0;
+        match amount {
+            a if a == lookup => {
+                c = Some(ZERO_COMMITMENT_LOOKUP_TABLE[mid].1);
```
I think you are meant to return here, otherwise this will infinite loop.

As a fun exercise I made a "constant time" algorithm:
```rust
    let log = amount.checked_ilog10().unwrap(); // don't actually unwrap here.
    let div = 10_u64.pow(log);

    let most_significant_digit = amount / div;

    if most_significant_digit * div != amount {
        return H_PRECOMP.vartime_multiscalar_mul([Scalar::from(amount), Scalar::ONE]);
    }

    let index = (most_significant_digit + u64::from(log) * 10 - 1 * (u64::from(log) + 1)) as usize;
    
    ZERO_COMMITMENT_LOOKUP_TABLE[index].1
```

note: this is not actually constant time for cryptographic purposes. it would be good to see how that stacks up.


FWIW: I would rather use rusts build in binary search even if it is a little slower.

SyntheticBird45 replied:
I must say you nailed it. Here's the benchmark:

![constant_time](https://github.com/user-attachments/assets/dd1abfa5-74ec-4100-968e-07b4207d338a)

```
lookup_search/binary_search_manual
                        time:   [835.15 ns 839.11 ns 843.73 ns]
lookup_search/constant_time
                        time:   [871.62 ns 873.04 ns 874.56 ns]
```

The manual implementation of binary search is faster than constant time yet yours is clever (imo).

SyntheticBird45 replied:
Forget to push the end pointer fix on the manual binary search one:

![fixed](https://github.com/user-attachments/assets/c8dbe51d-5a13-4650-978b-c60442d048c3)

```
lookup_search/binary_search_std
                        time:   [1.2080 µs 1.2101 µs 1.2124 µs]
lookup_search/binary_search_manual
                        time:   [991.33 ns 993.50 ns 995.41 ns]
lookup_search/hashmap_search_std
                        time:   [1.3606 µs 1.3620 µs 1.3636 µs]
lookup_search/constant_time
                        time:   [864.30 ns 865.61 ns 867.13 ns]
```

[Benchmark code.](https://gist.github.com/SyntheticBird45/e97c045aca71c5a0380dba8cf52bd0c5)

I obviously advocate we use your constant time version with a generous documentation on top.

SyntheticBird45 replied:
constant-time version will be used
https://matrix.to/#/!zPLCnZSsyeFFxUiqUZ:monero.social/$IBmC-hk2AN7XdQwo3T7StkaFDGEeRd-7XBeZenwBUCk?via=monero.social&via=matrix.org&via=kernal.eu

Boog900 commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,96 @@
+
+    // Binary search in lookup table (O(log2(n)) instead of O(n))
+    let (mut start, mut end, mut c): (usize, usize, Option<EdwardsPoint>) =
+        (0, (ZERO_COMMITMENT_LOOKUP_TABLE.len() - 1) / 2, None);
```
Also here we shouldn't be using the mid point as the end.

SyntheticBird45 replied:
Thx tests are passing now

hinto-janai commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,93 @@
+    while start <= end {
+        let mid = (start + end) / 2;
+        let lookup = ZERO_COMMITMENT_LOOKUP_TABLE[mid].0;
+        match amount {
+            a if a == lookup => {
+                return ZERO_COMMITMENT_LOOKUP_TABLE[mid].1;
+            }
+            a if a < lookup => {
+                end = mid - 1;
+            }
+            _ => {
+                start = mid + 1;
+            }
+        }
```
I think `clippy::comparison_chain` wants this:
```suggestion
        match amount.cmp(&lookup) {
            Ordering::Equal => {
                return ZERO_COMMITMENT_LOOKUP_TABLE[mid].1;
            }
            Ordering::Less => {
                end = mid - 1;
            }
            Ordering::Greater => {
                start = mid + 1;
            }
        }
```

hinto-janai commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,93 @@
+};
+use monero_serai::generators::H;
+
+/// This is the decomposed amount table containing the mandatory Pre-RCT amounts. It is use to pre-compute 
+/// zero commitments at runtime.
```
```suggestion
/// This is the decomposed amount table containing the mandatory Pre-RCT amounts. It is used to pre-compute 
/// zero commitments at runtime.
```

hinto-janai commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,93 @@
+    10000000000000000000
+];
+
+/// Runtime initialized H generator.
```
```suggestion
/// Runtime initialized [`H`] generator.
```

hinto-janai commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,93 @@
+
+/// Runtime initialized zero commitment lookup table
+///
+/// ASSUMPTION: This function assume that the [`ZERO_COMMITMENT_DECOMPOSED_AMOUNT`]
+/// table is sorted.
```
```suggestion
/// # Invariant
/// This function assumes that the [`ZERO_COMMITMENT_DECOMPOSED_AMOUNT`]
/// table is sorted.
```

hinto-janai commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,93 @@
+        lookup_table
+    });
+
+/// This function compute the zero commitment given a specific amount.
```
```suggestion
/// This function computes the zero commitment given a specific amount.
```

hinto-janai commented on helper/Cargo.toml:
```diff
@@ -10,12 +10,13 @@ repository = "https://github.com/Cuprate/cuprate/tree/main/consensus"
 
 [features]
 # All features off by default.
-default   = []
+default   = ["std"]
```
Either the comment needs to be updated or `"std"` should be removed, if `crypto` needs `std` it should be `crypto = ["std", "..."]`.

Boog900 commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,100 @@
+/// Compute it otherwise.
+#[expect(clippy::cast_possible_truncation)]
+pub fn compute_zero_commitments(amount: u64) -> EdwardsPoint {
+    if amount == 0 {
+        return H_PRECOMP.vartime_multiscalar_mul([Scalar::from(amount), Scalar::from(1_u8)]);
+    }
+
+    // OPTIMIZATION: Unlike monerod which execute a linear search across its lookup
+    // table (O(n)). Cuprate is making use of an arithmetic based constant time
+    // version (O(1)). It has been benchmarked in both hit and miss scenarios against
+    // a binary search lookup (O(log2(n))). To understand the following algorithm it
+    // is important to observe the pattern that follows the values of
+    // [`ZERO_COMMITMENT_DECOMPOSED_AMOUNT`].
+
+    // First obtain the logarithm base 10 of the amount. and extend it back to obtain
+    // the amount without its most significant digit.
+    let log = amount.checked_ilog10().unwrap();
```
```suggestion
    // OPTIMIZATION: Unlike monerod which execute a linear search across its lookup
    // table (O(n)). Cuprate is making use of an arithmetic based constant time
    // version (O(1)). It has been benchmarked in both hit and miss scenarios against
    // a binary search lookup (O(log2(n))). To understand the following algorithm it
    // is important to observe the pattern that follows the values of
    // [`ZERO_COMMITMENT_DECOMPOSED_AMOUNT`].

    // First obtain the logarithm base 10 of the amount. and extend it back to obtain
    // the amount without its most significant digit.
    let Some(log) = amount.checked_ilog10() else {
         // amount = 0 so H component is 0. 
         return ED25519_BASEPOINT_POINT;
    }
```

SyntheticBird45 replied:
Changed.

Boog900 commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,100 @@
+        return H_PRECOMP.vartime_multiscalar_mul([Scalar::from(amount), Scalar::ONE]);
+    }
+
+    // Calculating the index back by progressing within the powers of 10.
+    let index = (most_significant_digit + u64::from(log) * 10 - (u64::from(log) + 1)) as usize;
```
This is probably clearer:

```suggestion
    // Calculating the index back by progressing within the powers of 10.
    // The index of the first value in the cached amount's row.
    let row_start = u64::from(log) * 9;
    // The index of the cached amount
    let index = (most_significant_digit - 1 + row_start) as usize;
```

SyntheticBird45 replied:
Added

Boog900 commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,100 @@
+/// # Invariant
+/// This function assumes that the [`ZERO_COMMITMENT_DECOMPOSED_AMOUNT`]
+/// table is sorted.
+pub static ZERO_COMMITMENT_LOOKUP_TABLE: LazyLock<[(u64, EdwardsPoint); 172]> =
```
I don't think we need to keep the amount (`u64`) anymore.

SyntheticBird45 replied:
Removed

Boog900 reviewed (changes requested):
It would be good if you could add a test looping trough all the amounts in the table making sure we give the same commitment back from `compute_zero_commitment`

Boog900 commented on helper/src/crypto.rs:
```diff
@@ -0,0 +1,100 @@
+/// It will first attempt to lookup into the table of known Pre-RCT value.
+/// Compute it otherwise.
+#[expect(clippy::cast_possible_truncation)]
+pub fn compute_zero_commitments(amount: u64) -> EdwardsPoint {
```
nit:
```suggestion
pub fn compute_zero_commitment(amount: u64) -> EdwardsPoint {
```

SyntheticBird45 replied:
Fixed
